File: backend/app/runner/dynamic_functions.py
"""
Dynamic functions for workflow variable substitution
Provides functions like randomString(), randomNumber(), uuid(), timestamp(), etc.
that can be used in {{functionName(params)}} placeholders
"""
import random
import string
import uuid
import time
from datetime import datetime, timedelta
from typing import Any, Dict
import logging

logger = logging.getLogger(__name__)


class DynamicFunctions:
    """Collection of dynamic functions for variable substitution"""
    
    @staticmethod
    def randomString(length: int = 10) -> str:
        """
        Generate a random alphanumeric string
        
        Args:
            length: Length of the random string (default: 10)
            
        Returns:
            Random string of specified length
            
        Example: {{randomString(10)}} -> "aBcD1eFg2H"
        """
        try:
            length = int(length)
            if length < 1:
                length = 10
            return ''.join(random.choices(string.ascii_letters + string.digits, k=length))
        except (ValueError, TypeError):
            logger.warning(f"Invalid length for randomString: {length}, using default 10")
            return ''.join(random.choices(string.ascii_letters + string.digits, k=10))
    
    @staticmethod
    def randomNumber(size: int = 6) -> str:
        """
        Generate a random number
        
        Args:
            size: Number of digits (default: 6)
            
        Returns:
            Random number as string
            
        Example: {{randomNumber(6)}} -> "123456"
        """
        try:
            size = int(size)
            if size < 1:
                size = 6
            max_num = int('9' * size)
            min_num = int('1' + '0' * (size - 1)) if size > 1 else 1
            return str(random.randint(min_num, max_num))
        except (ValueError, TypeError):
            logger.warning(f"Invalid size for randomNumber: {size}, using default 6")
            return str(random.randint(100000, 999999))
    
    @staticmethod
    def randomEmail() -> str:
        """
        Generate a random email address
        
        Returns:
            Random email in format: random_string@example.com
            
        Example: {{randomEmail()}} -> "aBcD1eFg2H@example.com"
        """
        random_part = DynamicFunctions.randomString(10)
        return f"{random_part}@example.com"
    
    @staticmethod
    def uuid() -> str:
        """
        Generate a UUID v4
        
        Returns:
            UUID string
            
        Example: {{uuid()}} -> "550e8400-e29b-41d4-a716-446655440000"
        """
        return str(uuid.uuid4())
    
    @staticmethod
    def timestamp() -> str:
        """
        Get current Unix timestamp
        
        Returns:
            Current Unix timestamp as string
            
        Example: {{timestamp()}} -> "1730290800"
        """
        return str(int(time.time()))
    
    @staticmethod
    def iso_timestamp() -> str:
        """
        Get current ISO 8601 timestamp
        
        Returns:
            Current timestamp in ISO format
            
        Example: {{iso_timestamp()}} -> "2025-10-30T12:00:00"
        """
        return datetime.now().isoformat()
    
    @staticmethod
    def date(format: str = "%Y-%m-%d") -> str:
        """
        Get current date in specified format
        
        Args:
            format: Date format string (default: YYYY-MM-DD)
            
        Returns:
            Formatted date string
            
        Example: {{date()}} -> "2025-10-30"
        Example: {{date(%d/%m/%Y)}} -> "30/10/2025"
        """
        try:
            return datetime.now().strftime(format)
        except (ValueError, TypeError):
            logger.warning(f"Invalid date format: {format}, using default YYYY-MM-DD")
            return datetime.now().strftime("%Y-%m-%d")
    
    @staticmethod
    def futureDate(days: int = 1, format: str = "%Y-%m-%d") -> str:
        """
        Get a future date
        
        Args:
            days: Number of days in the future (default: 1)
            format: Date format string (default: YYYY-MM-DD)
            
        Returns:
            Formatted future date string
            
        Example: {{futureDate(7)}} -> "2025-11-06"
        """
        try:
            days = int(days)
            future = datetime.now() + timedelta(days=days)
            return future.strftime(format)
        except (ValueError, TypeError):
            logger.warning(f"Invalid parameters for futureDate: days={days}, format={format}")
            future = datetime.now() + timedelta(days=1)
            return future.strftime("%Y-%m-%d")
    
    @staticmethod
    def pastDate(days: int = 1, format: str = "%Y-%m-%d") -> str:
        """
        Get a past date
        
        Args:
            days: Number of days in the past (default: 1)
            format: Date format string (default: YYYY-MM-DD)
            
        Returns:
            Formatted past date string
            
        Example: {{pastDate(7)}} -> "2025-10-23"
        """
        try:
            days = int(days)
            past = datetime.now() - timedelta(days=days)
            return past.strftime(format)
        except (ValueError, TypeError):
            logger.warning(f"Invalid parameters for pastDate: days={days}, format={format}")
            past = datetime.now() - timedelta(days=1)
            return past.strftime("%Y-%m-%d")
    
    @staticmethod
    def randomChoice(options: str) -> str:
        """
        Pick a random choice from comma-separated options
        
        Args:
            options: Comma-separated options
            
        Returns:
            One randomly selected option
            
        Example: {{randomChoice(option1,option2,option3)}} -> "option2"
        """
        try:
            opts = [opt.strip() for opt in options.split(',')]
            if opts:
                return random.choice(opts)
            return ""
        except Exception as e:
            logger.warning(f"Error in randomChoice: {e}")
            return ""
    
    @staticmethod
    def randomAlpha(length: int = 10) -> str:
        """
        Generate a random alphabetic string (letters only, no numbers)
        
        Args:
            length: Length of the random string (default: 10)
            
        Returns:
            Random alphabetic string
            
        Example: {{randomAlpha(8)}} -> "aBcDeFgH"
        """
        try:
            length = int(length)
            if length < 1:
                length = 10
            return ''.join(random.choices(string.ascii_letters, k=length))
        except (ValueError, TypeError):
            logger.warning(f"Invalid length for randomAlpha: {length}, using default 10")
            return ''.join(random.choices(string.ascii_letters, k=10))
    
    @staticmethod
    def randomNumeric(length: int = 10) -> str:
        """
        Generate a random numeric string (digits only)
        
        Args:
            length: Length of the numeric string (default: 10)
            
        Returns:
            Random numeric string
            
        Example: {{randomNumeric(6)}} -> "123456"
        """
        try:
            length = int(length)
            if length < 1:
                length = 10
            return ''.join(random.choices(string.digits, k=length))
        except (ValueError, TypeError):
            logger.warning(f"Invalid length for randomNumeric: {length}, using default 10")
            return ''.join(random.choices(string.digits, k=10))
    
    @staticmethod
    def randomHex(length: int = 16) -> str:
        """
        Generate a random hexadecimal string
        
        Args:
            length: Length of the hex string (default: 16)
            
        Returns:
            Random hexadecimal string
            
        Example: {{randomHex(8)}} -> "a1b2c3d4"
        """
        try:
            length = int(length)
            if length < 1:
                length = 16
            return ''.join(random.choices(string.hexdigits[:16], k=length)).lower()
        except (ValueError, TypeError):
            logger.warning(f"Invalid length for randomHex: {length}, using default 16")
            return ''.join(random.choices(string.hexdigits[:16], k=16)).lower()
    
    @staticmethod
    def get_function(name: str):
        """Get a function by name"""
        functions = {
            'randomString': DynamicFunctions.randomString,
            'randomNumber': DynamicFunctions.randomNumber,
            'randomEmail': DynamicFunctions.randomEmail,
            'uuid': DynamicFunctions.uuid,
            'timestamp': DynamicFunctions.timestamp,
            'iso_timestamp': DynamicFunctions.iso_timestamp,
            'date': DynamicFunctions.date,
            'futureDate': DynamicFunctions.futureDate,
            'pastDate': DynamicFunctions.pastDate,
            'randomChoice': DynamicFunctions.randomChoice,
            'randomAlpha': DynamicFunctions.randomAlpha,
            'randomNumeric': DynamicFunctions.randomNumeric,
            'randomHex': DynamicFunctions.randomHex,
        }
        return functions.get(name)
    
    @staticmethod
    def get_all_functions() -> Dict[str, str]:
        """Get documentation for all available functions"""
        return {
            'randomString(length)': 'Generate a random alphanumeric string. Default length: 10',
            'randomNumber(size)': 'Generate a random number with specified digits. Default: 6 digits',
            'randomEmail()': 'Generate a random email address',
            'uuid()': 'Generate a UUID v4',
            'timestamp()': 'Get current Unix timestamp',
            'iso_timestamp()': 'Get current ISO 8601 timestamp',
            'date(format)': 'Get current date. Default format: %Y-%m-%d',
            'futureDate(days, format)': 'Get a future date. Default: 1 day, format: %Y-%m-%d',
            'pastDate(days, format)': 'Get a past date. Default: 1 day, format: %Y-%m-%d',
            'randomChoice(options)': 'Pick random choice from comma-separated options',
            'randomAlpha(length)': 'Generate random alphabetic string (letters only). Default: 10',
            'randomNumeric(length)': 'Generate random numeric string (digits only). Default: 10',
            'randomHex(length)': 'Generate random hexadecimal string. Default: 16',
        }

File: app/core/runner/dynamic_functions.ts
import type { ClockProvider, RngProvider } from "./harness/providers"

/**
 * Dynamic functions for workflow variable substitution.
 *
 * Ported from `backend/app/runner/dynamic_functions.py`. Every function goes
 * through the injected `ClockProvider` / `RngProvider` so the parity harness
 * can pin outputs deterministically (FixedClockProvider + SeededRandomProvider).
 * Production wires `WallClockProvider` + `CryptoRandomProvider`.
 *
 * Surface mirrors Python exactly (13 functions). No new functions added.
 */
export class DynamicFunctions {
  public constructor(
    private readonly clock: ClockProvider,
    private readonly rng: RngProvider,
  ) {}

  /** Random alphanumeric string (a-zA-Z0-9). Default length 10. */
  public randomString(length: unknown): string {
    const len = coerceLength(length, 10)
    const alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
    return sampleAlphabet(this.rng, alphabet, len)
  }

  /** Random number as a string with `size` digits. Default 6. */
  public randomNumber(size: unknown): string {
    const sz = coerceLength(size, 6)
    const max = Number("9".repeat(sz))
    const min = sz > 1 ? Number("1" + "0".repeat(sz - 1)) : 1
    const n = Math.floor(this.rng.next() * (max - min + 1)) + min
    return String(n)
  }

  /** Random email: `<randomString(10)>@example.com`. */
  public randomEmail(): string {
    return `${this.randomString(10)}@example.com`
  }

  /** UUID v4 string. */
  public uuid(): string {
    const bytes = this.rng.bytes(16)
    bytes[6] = (bytes[6]! & 0x0f) | 0x40
    bytes[8] = (bytes[8]! & 0x3f) | 0x80
    const hex = Buffer.from(bytes).toString("hex")
    return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`
  }

  /** Current Unix timestamp as a string. */
  public timestamp(): string {
    return String(Math.floor(this.clock.now().getTime() / 1000))
  }

  /** Current ISO 8601 timestamp. */
  public iso_timestamp(): string {
    return this.clock.now().toISOString()
  }

  /** Current date formatted. Default `%Y-%m-%d`. */
  public date(format?: unknown): string {
    return formatDate(this.clock.now(), typeof format === "string" && format.length > 0 ? format : "%Y-%m-%d")
  }

  /** Future date `days` from now. Default 1 day, `%Y-%m-%d`. */
  public futureDate(days?: unknown, format?: unknown): string {
    const d = coerceLength(days, 1)
    const fmt = typeof format === "string" && format.length > 0 ? format : "%Y-%m-%d"
    return formatDate(new Date(this.clock.now().getTime() + d * 86_400_000), fmt)
  }

  /** Past date `days` before now. Default 1 day, `%Y-%m-%d`. */
  public pastDate(days?: unknown, format?: unknown): string {
    const d = coerceLength(days, 1)
    const fmt = typeof format === "string" && format.length > 0 ? format : "%Y-%m-%d"
    return formatDate(new Date(this.clock.now().getTime() - d * 86_400_000), fmt)
  }

  /** Pick a random choice from comma-separated options. */
  public randomChoice(options: unknown): string {
    if (typeof options !== "string") return ""
    const opts = options.split(",").map((o) => o.trim()).filter((o) => o.length > 0)
    if (opts.length === 0) return ""
    const idx = Math.floor(this.rng.next() * opts.length)
    return opts[idx] ?? ""
  }

  /** Random alphabetic string (a-zA-Z). Default length 10. */
  public randomAlpha(length: unknown): string {
    const len = coerceLength(length, 10)
    const alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
    return sampleAlphabet(this.rng, alphabet, len)
  }

  /** Random numeric string (0-9). Default length 10. */
  public randomNumeric(length: unknown): string {
    const len = coerceLength(length, 10)
    const alphabet = "0123456789"
    return sampleAlphabet(this.rng, alphabet, len)
  }

  /** Random hexadecimal string (0-9a-f). Default length 16. */
  public randomHex(length: unknown): string {
    const len = coerceLength(length, 16)
    const alphabet = "0123456789abcdef"
    return sampleAlphabet(this.rng, alphabet, len)
  }

  /** Get a function by name, or `undefined` if unknown. */
  public getFunction(name: string): ((...args: unknown[]) => string) | undefined {
    const map: Record<string, (...args: unknown[]) => string> = {
      randomString: (a) => this.randomString(a),
      randomNumber: (a) => this.randomNumber(a),
      randomEmail: () => this.randomEmail(),
      uuid: () => this.uuid(),
      timestamp: () => this.timestamp(),
      iso_timestamp: () => this.iso_timestamp(),
      date: (a) => this.date(a),
      futureDate: (a, b) => this.futureDate(a, b),
      pastDate: (a, b) => this.pastDate(a, b),
      randomChoice: (a) => this.randomChoice(a),
      randomAlpha: (a) => this.randomAlpha(a),
      randomNumeric: (a) => this.randomNumeric(a),
      randomHex: (a) => this.randomHex(a),
    }
    return map[name]
  }

  /** Documentation map for the UI panel. */
  public getAllFunctions(): Record<string, string> {
    return {
      "randomString(length)": "Generate a random alphanumeric string. Default length: 10",
      "randomNumber(size)": "Generate a random number with specified digits. Default: 6 digits",
      "randomEmail()": "Generate a random email address",
      "uuid()": "Generate a UUID v4",
      "timestamp()": "Get current Unix timestamp",
      "iso_timestamp()": "Get current ISO 8601 timestamp",
      "date(format)": "Get current date. Default format: %Y-%m-%d",
      "futureDate(days, format)": "Get a future date. Default: 1 day, format: %Y-%m-%d",
      "pastDate(days, format)": "Get a past date. Default: 1 day, format: %Y-%m-%d",
      "randomChoice(options)": "Pick random choice from comma-separated options",
      "randomAlpha(length)": "Generate random alphabetic string (letters only). Default: 10",
      "randomNumeric(length)": "Generate random numeric string (digits only). Default: 10",
      "randomHex(length)": "Generate random hexadecimal string. Default: 16",
    }
  }
}

function coerceLength(value: unknown, fallback: number): number {
  const n = typeof value === "number" ? value : typeof value === "string" ? Number.parseInt(value, 10) : NaN
  if (!Number.isFinite(n) || n < 1) return fallback
  return Math.floor(n)
}

function sampleAlphabet(rng: RngProvider, alphabet: string, length: number): string {
  let out = ""
  for (let i = 0; i < length; i++) {
    const idx = Math.floor(rng.next() * alphabet.length)
    out += alphabet[idx] ?? ""
  }
  return out
}

/**
 * Minimal Python `strftime` subset covering the tokens Python's default
 * formats use (%Y-%m-%d, %d/%m/%Y, etc.) plus the common ones the UI exposes.
 * Anything unrecognized falls back to the literal char pair, matching the
 * Python behavior of `strftime` for unknown directives (platform-dependent,
 * but the parity harness only uses the tokens below).
 */
function formatDate(date: Date, format: string): string {
  const pad = (n: number, w = 2): string => String(n).padStart(w, "0")
  const replacements: Record<string, string> = {
    "%Y": String(date.getFullYear()),
    "%m": pad(date.getMonth() + 1),
    "%d": pad(date.getDate()),
    "%H": pad(date.getHours()),
    "%M": pad(date.getMinutes()),
    "%S": pad(date.getSeconds()),
    "%y": pad(date.getFullYear() % 100),
    "%B": MONTHS[date.getMonth()] ?? "",
    "%b": (MONTHS[date.getMonth()] ?? "").slice(0, 3),
    "%A": DAYS[date.getDay()] ?? "",
    "%a": (DAYS[date.getDay()] ?? "").slice(0, 3),
    "%%": "%",
  }
  let out = ""
  for (let i = 0; i < format.length; i++) {
    const ch = format[i]!
    if (ch === "%" && i + 1 < format.length) {
      const token = format.slice(i, i + 2)
      out += replacements[token] ?? token
      i += 1
    } else {
      out += ch
    }
  }
  return out
}

const MONTHS = [
  "January", "February", "March", "April", "May", "June",
  "July", "August", "September", "October", "November", "December",
]
const DAYS = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"]
